crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -902,8 +907,39 @@ impl CumulocityConverter {
             _ => {
                 // if device is unregistered register using auto-registration
                 if self.entity_store.get(&source).is_none() {
-                    let auto_registration_messages =
-                        self.entity_store.auto_register_entity(&source)?;
+                    let auto_registration_messages = match self
+                        .entity_store
+                        .auto_register_entity(&source)
+                    {
+                        Ok(auto_registration_messages) => auto_registration_messages,
+                        Err(e) => match e {
+                            entity_store::Error::NonDefaultTopicScheme(eid) => {
+                                // if its non supported enity, then cache it, log
+                                // and also send out error message on `errors` topic only at the first time.
+                                if !self.custom_entities.contains(&eid) {
+                                    if self.custom_entities.len() >= CUSTOM_ENTITY_CACHE_SIZE {
+                                        self.custom_entities.pop_front();
+                                    }
+                                    self.custom_entities.push_back(eid.clone());
```

Comment:
To avoid this code complexity at this level, I'd create a `RingBuffer` struct which wraps the `VecDeque` internally and move this logic into its `push` function.

Reply:
Not relevant anymore